vignettes/invariants.Rmd

```diff
@@ -190,38 +187,50 @@ df2[[2]]
 .subset2(df2, 2)
 ```
 
-### Corollary
+NB: `x[[j]]` always returns an object of size `nrow(x)` if the column exists.
 
-Accessing `NA` or missing columns returns `NULL`.
-Out-of-bounds indexing throws an error.
+`j` must be a single number or a string, as enforced by `vec_as_index(j)`.
+
+```{r double-bracket-requires-scalar-j-index, dftbl = TRUE}
+df[[1:2]]
+df[[c("n", "c")]]
+df[[TRUE]]
+df[[mean]]
+```
+
+<!-- HW: the failure with mean needs a better error. -->
+
+Missing and out-of-bounds (OOB) values of `j` return `NULL` with a warning or error:
 
 ```{r double-bracket-j-na, dftbl = TRUE}
-df[[NA]]
-df[[NA_integer_]]
 df[[NA_character_]]
 df[["x"]]
+
+df[[NA_integer_]]
+df[[-1]]
 df[[4]]
 ```
 
-`[[` requires scalar `vec_as_index(j)`.
+<!-- HW: why is x[[NA]] an OOB error and not a type error? -->
```

Comment:
Can we review and prioritize error messages after we're done defining behavior? I haven't paid too much attention yet, some messages aren't defined yet (instead an error of the type "function `error_...()` not found" is raised).

Reply:
Ok, I'll just leave these comments where we need to refine behaviour to match invariants.